Init Library page

pva701 opened the pull request:
## Description

This issue only introduces a basic table for LibraryPage. As new UI mock up design arrived, it will be develop in further PR.

Also, preview table for UploadPage is added.

## Related issue(s)

<!--
- Short description of how the PR relates to the issue, including an issue link.
For example
- Fixed #100500 by adding lenses to exported items

Write 'None' if there are no related issues (which is discouraged).
-->

https://issues.serokell.io/issue/EDNA-31

## :white_check_mark: Checklist for your Pull Request

<!--
Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

If you don't set a checkmark (e. g. don't add a test for new functionality),
you must be able to justify that.
-->

#### Related changes (conditional)

- Tests
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [ ] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)

#### Stylistic guide (mandatory)

- [ ] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [ ] My code complies with the [style guide](../tree/master/docs/code-style.md).

pva701 committed:
Initialise library page. Add simple table

Problem: we need to show entities from library in the tables.

Solution: implement simple table component for it.

pva701 committed:
Add add button to entityTabs component

Problem: we need to stub button to add entities on library page.

Solution: add new styled button.

pva701 committed:
Generialise Table component. Reuse it on Upload page

Problem: we need to use the same styling for a table on Upload page.

Solution: generialise Library Table and use it.

sashasashasasha151 commented on frontend/src/components/SuspsenseSpinner.tsx:
```diff
@@ -0,0 +1,6 @@
+
+export function SuspenseSpinner({ children }: { children: React.ReactNode }) {
+  return <Suspense fallback={<div className="spinner" />}>{children}</Suspense>;
+}
```
What about put it in spinner folder?

sashasashasasha151 commented on frontend/src/components/SuspsenseSpinner.tsx:
```diff
@@ -0,0 +1,6 @@
+import "./Spinner.scss";
+
+export function SuspenseSpinner({ children }: { children: React.ReactNode }) {
+  return <Suspense fallback={<div className="spinner" />}>{children}</Suspense>;
```
Looks like experiment feature, do we really want it? (just a question, for real it looks cool)

pva701 replied:
yes, it's convenient that I don't have to have this 
`if (loadableContent.status === 'loading')` or something like that

sashasashasasha151 commented on frontend/src/components/Button/Button.scss:
```diff
@@ -1,5 +1,12 @@
 @import "src/styles/variables";
 
+@mixin btn-typography() {
```
`btn-typography` but only for two buttons. Still do not understand why you don't want to combine it in one class.

pva701 replied:
because fuck off

pva701 replied:
it's renamed to `snd-btn` and `main-btn` in future prs

sashasashasasha151 commented on frontend/src/components/Button/Button.scss:
```diff
@@ -1,5 +1,12 @@
 
+@mixin btn-typography() {
+  font-size: 14 * $px;
+  font-style: normal;
+  font-weight: normal;
```
Looks like it is normal by default

pva701 replied:
oh, I just forgot we are going  to nominate this project at the contest "The shortest code", and one extra line in favour of clarity is unacceptable

sashasashasasha151 commented on frontend/src/components/Button/Button.scss:
```diff
@@ -1,5 +1,12 @@
 @import "src/styles/variables";
 
+@mixin btn-typography() {
```
Please add `cursor pointer` for all buttons

pva701 replied:
it's added in further PR

sashasashasasha151 commented on frontend/src/components/Button/Button.scss:
```diff
@@ -49,24 +56,38 @@
-    background-color: #cbcbcb;
+    color: white;
+    background-color: $color-green;
+  }
```
You can just add `pointer-events: none` for disabled and do not write `:not(:disabled)`

sashasashasasha151 commented on frontend/src/components/Button/Button.scss:
```diff
@@ -49,24 +56,38 @@
+
+  padding: (9 * $px) (20 * $px);
+
+  color: #3a8370;
```
$color-green?

sashasashasasha151 commented on frontend/src/components/Button/Button.scss:
```diff
@@ -49,24 +56,38 @@
-  color: #303030;
-  background-color: #f2f2f2;
-  border-color: #c4c4c4;
-  border-radius: 8 * $px;
+  @include btn-typography();
```
No margin between `plus` and button text

pva701 replied:
because it looks ugly, not going to fix. it's from the category: fix/discuss with the designers when time comes

sashasashasasha151 replied:
>it looks ugly

But you do not create design

pva701 replied:
yes, but unlike you I don't treat this like: I am told to do and I have to do, I treat the project like I have option to negotiate with designers the design, and say: look these margins seem to wide, there it seems weird, and so on and so forth. So it's constant negotiation. And this is just minor one which I don't wan to spend time on right now because it's unessential for now.

sashasashasasha151 replied:
And I like this margin and because of this you need to ask designers first and do not use your opinion as a good point to do what you want (but ok)

pva701 replied:
No, you understand this's totally wrongly and in totally submissive way :putya:

In this position I leave this decision up to me: what to discuss and what to not. Actually I don't discuss anything, just point in my mind to ask someday/on next meeting. Because thankfully I am not you and I understand that if I asked about any such small detail, the development of MVP would take much more time, but underlying it's one more time: for MVP I don't make the goal to correspond totally. It would be ideally, but I am sure we won't get in time anyway, though, it's not the reason to make it if we have time.

Why I do so? Because one day designers will look over the design and point to all such things. If they don't, it would mean it's so insignificant that they don't even notice the difference, why then I should to spend time now?

If you perceive the mock up design as 100% goal to reach, I treat it as good to reach but 
1. on this stage it's ok to miss something
2. it's up to me to discuss with designers some weird things, and make sure they really insist on doing something.

I including sometimes I just don't do it in a wrong way because it's easier and because it seems to be as better/not essential to follow for now but will be discussed later

pva701 replied:
It's a way of treating: you treat it's like ok, I got the task I have to do. But I treat it like: ok I think the task should be this because of this reason and this, so I build the reality myself :putya:

pva701 replied:
I am sure you didn't even run the mock in presentation mode to look how it will look on your screen :putya: Because in the mock up, it looks: oh nice design, nice margin, but if you run this in presentation mode, you will see that 15px might be pretty big space.

pva701 replied:
Last example of this: I made horizontal lines (borders) in the table shorter, no as wide as screen size and told to Darya about my concerns today. And you know what? She agreed that it was perhaps not the best decision. and such things, when we discuss something happens constantly under the hood, but you still seem to think the mock up is high quality undeniable decision to follow :putya:

sashasashasasha151 commented on frontend/src/components/UploadPreviewTable/UploadPreviewTable.tsx:
```diff
@@ -1,45 +1,67 @@
+    ex.compounds.forEach(compound => {
+      data.push({
+        compound,
+        target: ex.target + (ex.isNew ? "*" : ""),
```
Why not template strings?

sashasashasasha151 commented on frontend/src/components/UploadPreviewTable/UploadPreviewTable.tsx:
```diff
@@ -1,45 +1,67 @@
-          <th />
-        </tr>
-      </thead>
+  const previewColumns = React.useMemo(
```
What for do you need `useMemo` here?

pva701 replied:
https://react-table.tanstack.com/docs/quick-start#getting-your-data

sashasashasasha151 commented on frontend/src/components/UploadPreviewTable/UploadPreviewTable.tsx:
```diff
@@ -1,45 +1,67 @@
-        )}
-      </tbody>
-    </table>
+  const data: PreviewRow[] = [];
+
+  targets.map(ex =>
```
Why not put them in one function?

pva701 replied:
put what?

sashasashasasha151 commented on frontend/src/components/UploadPreviewTable/UploadPreviewTable.scss:
```diff
@@ -1,19 +1,17 @@
-  }
+.uploadPreviewTable {
+  width: 100%;
+  border: $px solid #9ba5a1;
```
As I can see the color is wrong

pva701 replied:
what? it's taken directly from Figma design

sashasashasasha151 replied:
![image](https://user-images.githubusercontent.com/25207759/111996718-5cb2ac00-8b2b-11eb-9f1b-f5ed7e335cb0.png)
Them wrong border color on other elements (it is from second pr)

pva701 replied:
other borders will be fixed in follow-ups of EDNA-31. as you can see there are more even discrepancies with the Figma mock up, but preview table table is going to be done within this PR because I focused on it in this PR.

sashasashasasha151 commented on frontend/src/components/UploadPreviewTable/UploadPreviewTable.scss:
```diff
@@ -1,19 +1,17 @@
-.uploadPreviewTable {
+.uploadPreviewTableContainer {
   max-height: 300 * $px;
+  overflow-y: scroll;
```
![image](https://user-images.githubusercontent.com/25207759/111929428-5a1e6b00-8ac7-11eb-8af5-d1cb92eaf1eb.png)
Looks like not what we need from scrolling

pva701 replied:
![image](https://user-images.githubusercontent.com/7355273/111988042-080a3380-8b21-11eb-9830-b5e2d685d814.png)
better?

sashasashasasha151 commented on frontend/src/components/Table/Table.scss:
```diff
@@ -0,0 +1,83 @@
+
+  &__columnHead {
+    position: relative;
+    padding-top: $head-vertical-padding;
+    padding-bottom: $head-vertical-padding;
+    padding-left: $cell-indent;
```
Just padding

pva701 replied:
no, it's indent.

sashasashasasha151 replied:
I mean you can write one line instead of three

sashasashasasha151 commented on frontend/src/components/Table/Table.scss:
```diff
@@ -0,0 +1,83 @@
+    padding-left: $cell-indent;
+
+    font-size: 14 * $px;
+    font-style: normal;
```
No need to write it

sashasashasasha151 commented on frontend/src/components/Table/Table.scss:
```diff
@@ -0,0 +1,83 @@
+
+  &__headRightBorder {
+    $height: 20 * $px;
+    position: absolute;
```
I think it is possible to just use flex on the parent element

pva701 replied:
We can think a lot of different things.... didn't work out well for me, you can try yourself and say me why as you have your glorious book now :putya:

sashasashasasha151 commented on frontend/src/components/Table/Table.scss:
```diff
@@ -0,0 +1,83 @@
+    background-color: #eaeaea;
+  }
+
+  &__rowHovered {
+    &:hover {
+      background-color: $color-light-green;
+    }
+  }
```
Use element + modifier instead of combining them

pva701 replied:
what? ~sounds like time wasting bullshit which nobody even cares, which you just can't keep for yourself and can't refrain from feeding you importance feeling telling that~

sashasashasasha151 replied:
Sounds like good structure and better understanding of the code instead of writing things which only you can understand

pva701 replied:
can you provide statistical data that this is so not-understandable that only I can understand. I doubts that it's not from your reality, where bullshit like this has maximum priotity.

but in general I don't understand what you want from me here?

pva701 replied:
I just use standard state on hover to change some attribute, how you suggest changing it?

sashasashasasha151 commented on frontend/src/components/Table/Table.scss:
```diff
@@ -0,0 +1,83 @@
+    padding-top: 16 * $px;
+    padding-bottom: 16 * $px;
+
+    padding-left: $cell-indent;
```
Just padding

sashasashasasha151 commented on frontend/src/components/Table/Table.scss:
```diff
@@ -0,0 +1,83 @@
+    border-bottom: $px solid $border-color;
+  }
+
+  &__cell_odd {
```
Move modifiers to the cell

pva701 replied:
done

sashasashasasha151 commented on frontend/src/components/Table/Table.tsx:
```diff
@@ -0,0 +1,83 @@
+    data,
+  });
+
+  let lastColumnWithRightBorder = columns.length - 2;
+  for (let i = columns.length - 1; i >= 0; i--) {
+    lastColumnWithRightBorder = i - 1;
+    if (columns[i].Header) {
+      break;
+    }
+  }
```
Can you please explain what is it and what for?

pva701 replied:
to remove this tiny | for empty headers. so if there are several empty headers at the end, none of them will have |

sashasashasasha151 replied:
Move to function?

pva701 replied:
ok

sashasashasasha151 commented on frontend/src/pages/library/LibraryPage.tsx:
```diff
@@ -1,6 +1,120 @@
-  return <PageLayout>Library Page</PageLayout>;
+  return (
+    <PageLayout>
+      <EntitiesTab
```
Looks like over engineering but I do not want to argue. Just remove copy paste

pva701 replied:
>over engineering

>remove copy paste

it was introduced exactly to reduce the copy-paste. this copy-paste is eliminated in further PRs

pva701 committed:
Make headers of preview table fixed. Make content scrollable.

Problem: it doesn't look so good, also borders are hidden under scrolling.

Solution: fix it, also fix other minor issues.

pva701 committed:
Merge branch 'master' into pva/init-library-page
